executor.go

```diff
@@ -777,14 +783,301 @@ func (e *executor) executeDifferenceShard(ctx context.Context, index string, c *
+	Count uint64     `json:"count"`
+}
+
+// mergeGroupCounts merges two slices of GroupCounts throwing away any that go
+// beyond the limit. It assume that the two slices are sorted by the row ids in
+// the fields of the group counts. It may modify its arguments.
+func mergeGroupCounts(a, b []GroupCount, limit int) []GroupCount {
+	if limit > len(a)+len(b) {
+		limit = len(a) + len(b)
+	}
+	ret := make([]GroupCount, 0, limit)
+	i, j := 0, 0
+	for i < len(a) && j < len(b) && len(ret) < limit {
+		switch a[i].Compare(b[j]) {
+		case -1:
+			ret = append(ret, a[i])
+			i++
+		case 0:
+			a[i].Count += b[j].Count
+			ret = append(ret, a[i])
+			i++
+			j++
+		case 1:
+			ret = append(ret, b[j])
+			j++
+		}
+	}
+	for ; i < len(a) && len(ret) < limit; i++ {
+		ret = append(ret, a[i])
+	}
+	for ; j < len(b) && len(ret) < limit; j++ {
+		ret = append(ret, b[j])
+	}
+	return ret
+}
+
+func (g GroupCount) Compare(o GroupCount) int {
+	for i := range g.Group {
+		if g.Group[i].RowID < o.Group[i].RowID {
+			return -1
+		}
+		if g.Group[i].RowID > o.Group[i].RowID {
+			return 1
+		}
+	}
+	return 0
+}
+
+func (e *executor) executeGroupByShard(_ context.Context, index string, c *pql.Call, shard uint64, childRows []RowIDs) ([]GroupCount, error) {
+	iter := newGroupByIterator(childRows, c.Children, index, shard, e.Holder)
+	if iter == nil {
+		return []GroupCount{}, nil
+	}
+
+	limit := int(^uint(0) >> 1)
+	if lim, hasLimit, err := c.UintArg("limit"); err != nil {
+		return nil, err
+	} else if hasLimit {
+		limit = int(lim)
+	}
+
+	results := make([]GroupCount, 0)
+
+	num := 0
+	for gc, done := iter.Next(); !done && num < limit; gc, done = iter.Next() {
+		if gc.Count > 0 {
+			num++
+			results = append(results, gc)
+		}
+	}
+
+	return results, nil
+}
+
+func (e *executor) executeRows(ctx context.Context, index string, c *pql.Call, shards []uint64, opt *execOptions) (RowIDs, error) {
+	if columnID, ok, err := c.UintArg("column"); err != nil {
+		return nil, errors.Wrap(err, "getting column")
+	} else if ok {
+		shards = []uint64{columnID / ShardWidth}
+	}
+
+	// Execute calls in bulk on each remote node and merge.
+	mapFn := func(shard uint64) (interface{}, error) {
+		return e.executeRowsShard(ctx, index, c, shard)
+	}
+
+	// Determine limit so we can use it when reducing.
+	limit := int(^uint(0) >> 1)
+	if lim, hasLimit, err := c.UintArg("limit"); err != nil {
+		return nil, err
+	} else if hasLimit {
+		limit = int(lim)
+	}
+
+	// Merge returned results at coordinating node.
+	reduceFn := func(prev, v interface{}) interface{} {
+		other, _ := prev.(RowIDs)
+		return other.merge(v.(RowIDs), limit)
+	}
+	// Get full result set.
+	other, err := e.mapReduce(ctx, index, shards, c, opt, mapFn, reduceFn)
+	if err != nil {
+		return nil, err
+	}
+	results, _ := other.(RowIDs)
+	return results, nil
+}
+
+func (e *executor) executeRowsShard(_ context.Context, index string, c *pql.Call, shard uint64) (RowIDs, error) {
+	// Fetch index.
```

Comment:
agree - could cause confusion otherwise. Might be worth rethinking at some point though. I wonder if we could work the index not found error into the field not found error.